### PR TITLE
Add `cupy.roots` for Hermitian or symmetric matrix

### PR DESCRIPTION
**Pull request by Dahlia-Chehata**

### CODE REVIEW

File: tests/cupy_tests/lib_tests/test_polynomial.py [hunk 1]
```diff
@@ -8,6 +9,15 @@
 from cupy import testing
 
 
+def _wraps_root_test(func):
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        out = func(*args, **kwargs)
+        out.sort()
```
Comment (Dahlia-Chehata) [inline tests/cupy_tests/lib_tests/test_polynomial.py:16]: The current implementation doesn't guarantee Numpy's order of the output

File: tests/cupy_tests/lib_tests/test_polynomial.py [hunk 2]
```diff
@@ -403,3 +425,92 @@ def test_polyadd_diff_types_poly1d(self, xp, dtype1, dtype2):
...
+        a = xp.array([3j, 1.5j, -3j], dtype)
+        return xp.roots(a)
+
+    @testing.for_all_dtypes(no_float16=True, no_bool=True)
+    @testing.numpy_cupy_array_equal()
+    def test_roots_two_sized1(self, xp, dtype):
+        a = xp.array([5, 10], dtype)
+        return xp.roots(a)
+
+    @testing.for_all_dtypes(no_bool=True)
```
Comment (Dahlia-Chehata) [inline tests/cupy_tests/lib_tests/test_polynomial.py:460]: bool tests are being skipped because of as_series. I think trying a workaround to support them isn't necessary since roots should be numerical values ?

File: tests/cupy_tests/lib_tests/test_polynomial.py [hunk 3]
```diff
@@ -22,7 +32,7 @@ def test_poly1d_numpy_array(self, xp, dtype):
         with cupyx.allow_synchronize(False):
             out = xp.poly1d(a, variable=self.variable)
         assert out.variable == (self.variable or 'x')
-        return out.coeffs
```
Comment (asi1024) [inline tests/cupy_tests/lib_tests/test_polynomial.py:25]: This fix is unrelated to this PR. Please split into another PR if you want to work on the refactoring.

File: tests/cupy_tests/lib_tests/test_polynomial.py [hunk 4]
```diff
@@ -125,6 +134,19 @@ def test_poly1d_order_leading_zeros(self, xp, dtype):
         a = xp.array([0, 0, 1, 2, 3, 0], dtype)
         return xp.poly1d(a).order
 
+    @testing.for_signed_dtypes()
+    @testing.numpy_cupy_allclose(rtol=1e-6)
+    @_wraps_root_test
+    def test_poly1d_roots1(self, xp, dtype):
+        a = xp.array([-3, -2.5, 3], dtype)
+        return xp.poly1d(a).roots
```
Comment (asi1024) [inline tests/cupy_tests/lib_tests/test_polynomial.py:142]: Why not sort the return value directly for the readability?
```suggestion
    @testing.for_signed_dtypes()
    @testing.numpy_cupy_allclose(rtol=1e-6)
    def test_poly1d_roots1(self, xp, dtype):
        a = xp.array([-3, -2.5, 3], dtype)
        out = xp.poly1d(a).roots
        # The current `cupy.roots` doesn't guarantee the order of results.
        return xp.sort(out)
```

File: cupy/lib/_routines_poly.py [hunk 5]
```diff
@@ -71,3 +71,43 @@ def polysub(a1, a2):
...
+        return cupy.array([])
+    [p] = cupy.polynomial.polyutils.as_series([p[::-1]])
+    if p.size < 2:
+        return cupy.array([])
+    if p.size == 2:
+        out = cupy.array([-p[0]/p[1]])
+        if p[0] == 0:
+            out = out.real.astype(float)
+        return out
+    cmatrix = cupy.polynomial.polynomial.polycompanion(p)
```
Comment (asi1024) [inline cupy/lib/_routines_poly.py:106]: Are they fast path for small cases, or workaround?

Comment (Dahlia-Chehata) [reply cupy/lib/_routines_poly.py:106]: workaround

File: tests/cupy_tests/lib_tests/test_polynomial.py (hunk 2 above)
Comment (asi1024) [reply tests/cupy_tests/lib_tests/test_polynomial.py:472]: How about raising `NotImplementedError` for boolean inputs?

File: cupy/lib/_routines_poly.py [hunk 6]
```diff
@@ -71,3 +71,43 @@ def polysub(a1, a2):
...
+    """
+    if p.ndim == 0:
+        raise TypeError('0-dimensional input is not allowed')
+    if p.size < 2 or (p.ndim == 1 and not cupy.any(p)):
+        return cupy.array([])
+    [p] = cupy.polynomial.polyutils.as_series([p[::-1]])
+    if p.size < 2:
+        return cupy.array([])
+    if p.size == 2:
+        out = cupy.array([-p[0]/p[1]])
```
Comment (asi1024) [inline cupy/lib/_routines_poly.py:102]: Reduce memory copy.
```suggestion
        out = (-p[0] / p[1])[None]
```

File: cupy/lib/_routines_poly.py [hunk 7]
```diff
@@ -71,3 +71,43 @@ def polysub(a1, a2):
...
+        if p[0] == 0:
+            out = out.real.astype(float)
```
Comment (asi1024) [inline cupy/lib/_routines_poly.py:104]: For the readability.
```suggestion
            out = out.real.astype(numpy.float64)
```

File: cupy/lib/_routines_poly.py [hunk 8]
```diff
@@ -71,3 +71,43 @@ def polysub(a1, a2):
...
+        This function doesn't support currently polynomial coefficients
+        whose companion matrices are general 2d square arrays. Only those
+        with complex Hermitian or real symmetric 2d arrays are allowed.
+
+    .. seealso:: :func:`numpy.roots`
+
+    """
+    if p.ndim == 0:
+        raise TypeError('0-dimensional input is not allowed')
+    if p.size < 2 or (p.ndim == 1 and not cupy.any(p)):
```
Comment (asi1024) [inline cupy/lib/_routines_poly.py:96]: When `p.ndim == 1 and not cupy.any(p)`, the condition in `L99` seems evaluated to `True`.
```suggestion
    if p.size < 2:
```

File: tests/cupy_tests/lib_tests/test_polynomial.py [hunk 9]
```diff
@@ -125,6 +125,20 @@ def test_poly1d_order_leading_zeros(self, xp, dtype):
         a = xp.array([0, 0, 1, 2, 3, 0], dtype)
         return xp.poly1d(a).order
 
+    @testing.for_signed_dtypes()
+    @testing.numpy_cupy_allclose(rtol=1e-6)
+    def test_poly1d_roots1(self, xp, dtype):
+        a = xp.array([-3, -2.5, 3], dtype)
+        out = xp.poly1d(a).roots
+        # The current `cupy.roots` doesn't guarantee the order of results.
```
Comment (asi1024) [inline tests/cupy_tests/lib_tests/test_polynomial.py:133]: Could you write this comment also in the docstring of `cupy.roots`?

File: tests/cupy_tests/lib_tests/test_polynomial.py [hunk 10]
```diff
@@ -423,3 +437,102 @@ def test_polyroutine_diff_types_poly1d(self, xp, dtype1, dtype2):
...
         return out
+
+
+@testing.gpu
+class TestRoots(unittest.TestCase):
+
+    @testing.for_signed_dtypes()
+    @testing.numpy_cupy_allclose(rtol=1e-6)
+    def test_roots1(self, xp, dtype):
+        a = xp.array([2, -1, -2], dtype)
```
Comment (asi1024) [inline tests/cupy_tests/lib_tests/test_polynomial.py:448]: Why not parameterize inputs with `testing.parameterize`?

File: cupy/lib/_routines_poly.py [hunk 11]
```diff
@@ -71,3 +73,47 @@ def polysub(a1, a2):
...
+    if p.size == 2:
+        out = (-p[0] / p[1])[None]
+        if p[0] == 0:
+            out = out.real.astype(numpy.float64)
+        return out
+    cmatrix = cupy.polynomial.polynomial.polycompanion(p)
+    if cupy.array_equal(cmatrix, cmatrix.conj().T):
+        out = cupy.linalg.eigvalsh(cmatrix)
+    else:
+        raise NotImplementedError('Only complex Hermitian and real '
```
[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/lib/_routines_poly.py:117]: Please add in the else statement.

File: tests/cupy_tests/lib_tests/test_polynomial.py [hunk 12]
```diff
@@ -125,6 +125,20 @@ def test_poly1d_order_leading_zeros(self, xp, dtype):
         a = xp.array([0, 0, 1, 2, 3, 0], dtype)
         return xp.poly1d(a).order
 
+    @testing.for_signed_dtypes()
+    @testing.numpy_cupy_allclose(rtol=1e-6)
+    def test_poly1d_roots1(self, xp, dtype):
```
Comment (asi1024) [inline tests/cupy_tests/lib_tests/test_polynomial.py:130]: What is the role of this test? It looks able to be merged into `TestRootsReal`.

File: tests/cupy_tests/lib_tests/test_polynomial.py [hunk 13]
```diff
@@ -125,6 +125,20 @@ def test_poly1d_order_leading_zeros(self, xp, dtype):
...
+        a = xp.array([-3, -2.5, 3], dtype)
+        out = xp.poly1d(a).roots
+        # The current `cupy.roots` doesn't guarantee the order of results.
+        return xp.sort(out)
+
+    @testing.for_all_dtypes(no_bool=True)
+    def test_poly1d_roots2(self, dtype):
```
Comment (asi1024) [inline tests/cupy_tests/lib_tests/test_polynomial.py:137]: Duplicated with `TestRoots.test_roots_general`.

File: tests/cupy_tests/lib_tests/test_polynomial.py [hunk 14]
```diff
@@ -423,3 +437,91 @@ def test_polyroutine_diff_types_poly1d(self, xp, dtype1, dtype2):
...
+
+    @testing.for_all_dtypes(no_bool=True)
+    def test_roots_zeros2(self, dtype):
+        for xp in (numpy, cupy):
+            a = xp.zeros((2, 1), dtype)
+            with pytest.raises(ValueError):
+                cupy.roots(a)
+
+    def test_roots_bool(self):
+        a = testing.shaped_arange((5,), cupy, bool)
```
Comment (asi1024) [inline tests/cupy_tests/lib_tests/test_polynomial.py:525]: Please test with a symmetric matrix.

File: tests/cupy_tests/lib_tests/test_polynomial.py (hunk 12 above)
Comment (Dahlia-Chehata) [reply tests/cupy_tests/lib_tests/test_polynomial.py:130]: Please don't confuse `poly1d.roots` with `cupy.roots()`. I know it is simply one func calling the other but I thought this call should be tested too

File: tests/cupy_tests/lib_tests/test_polynomial.py (hunk 13 above)
Comment (Dahlia-Chehata) [reply tests/cupy_tests/lib_tests/test_polynomial.py:137]: Ditto

File: tests/cupy_tests/lib_tests/test_polynomial.py (hunk 12 above)
Comment (asi1024) [reply tests/cupy_tests/lib_tests/test_polynomial.py:130]: `.roots` is just an alias of `cupy.roots`, so we don't need to double-check of this routine except for the typecheck of the return value. Also the test of `poly1d.__add__` is only for typecheck of the return value, and more detailed tests, for example dtype combination, are checked only in `polyadd` test.

File: tests/cupy_tests/lib_tests/test_polynomial.py [hunk 15]
```diff
@@ -423,3 +437,96 @@ def test_polyroutine_diff_types_poly1d(self, xp, dtype1, dtype2):
...
+    def test_roots_zeros1(self, xp, dtype):
+        a = xp.zeros((3,), dtype)
+        return xp.roots(a)
+
+    @testing.for_all_dtypes(no_bool=True)
+    def test_roots_zeros2(self, dtype):
+        for xp in (numpy, cupy):
+            a = xp.zeros((2, 1), dtype)
+            with pytest.raises(ValueError):
+                cupy.roots(a)
```
Comment (asi1024) [inline tests/cupy_tests/lib_tests/test_polynomial.py:522]:
```suggestion
                xp.roots(a)
```

File: tests/cupy_tests/lib_tests/test_polynomial.py [hunk 16]
```diff
@@ -423,3 +437,96 @@ def test_polyroutine_diff_types_poly1d(self, xp, dtype1, dtype2):
...
+            cupy.roots(a)
+
+    @testing.for_all_dtypes(no_bool=True)
+    @testing.numpy_cupy_array_equal()
+    def test_roots_zeros1(self, xp, dtype):
+        a = xp.zeros((3,), dtype)
+        return xp.roots(a)
+
+    @testing.for_all_dtypes(no_bool=True)
+    def test_roots_zeros2(self, dtype):
```
Comment (asi1024) [inline tests/cupy_tests/lib_tests/test_polynomial.py:518]: Rename the test clearer.
```suggestion
    def test_roots_zeros_ndim(self, dtype):
```

File: tests/cupy_tests/lib_tests/test_polynomial.py (hunk 12 above)
Comment (Dahlia-Chehata) [reply tests/cupy_tests/lib_tests/test_polynomial.py:130]: I know, I didn't repeat `cupy.roots` tests for `.roots` . It is only a single test (taking into consideration that general matrix related tests will be removed). Do you want to omit it?

File: cupy/lib/_routines_poly.py [hunk 17]
```diff
@@ -71,3 +73,48 @@ def polysub(a1, a2):
         out = out.astype(cupy.result_type(a1, a2), copy=False)
         out -= 2 * out - a1
     return out
+
+
+def roots(p):
+    """Computes the roots of a polynomial with given coefficients.
+
+    Args:
+        p (cupy.ndarray): polynomial coefficients.
```
Comment (asi1024) [inline cupy/lib/_routines_poly.py:82]: Why not support for poly1d inputs?

Comment (Dahlia-Chehata) [reply cupy/lib/_routines_poly.py:82]: pol1d inputs are not documented in numpy's documentation

Comment (asi1024) [reply cupy/lib/_routines_poly.py:82]: NumPy's documentation says that the parameters should be `rank-1 array`. So please support it.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM.
